### ISSUE TITLE
Is the `limitStr` function unintentionally counting ANSI control characters when truncating strings?

### ISSUE THREAD

**Issue by greglanthier** [Evertras/bubble-table#66]
Thank you for sharing **bubble-table** - its easy to use and full of great features.

I have a question about expected behaviour when it comes to how text is rendered in a cell.

I was trying to place a progress-bar using **github.com/charmbracelet/bubbles/progress** into a cell in a table using code that looks loosely like this:

```golang
	prog := progress.New(
		progress.WithScaledGradient("#FF7CCB", "#FDFF8C"),
		progress.WithWidth(18),
	)
	prog.Full = ':'
	prog.Empty = ' '
	m.table = table.New([]table.Column{
		table.NewFlexColumn("cpu", "CPU", 1),
	}).WithRows([]table.Row{
		table.NewRow(table.RowData{
			"cpu":      fmt.Sprintf("[%s]", prog.ViewAs(0.34)),
		}),
		table.NewRow(table.RowData{
			"cpu":      fmt.Sprintf("[%s]", prog.ViewAs(0.54)),
		}),
		table.NewRow(table.RowData{
			"cpu":      fmt.Sprintf("[%s]", prog.ViewAs(0.84)),
		}),
	}).WithBaseStyle(lipgloss.NewStyle().Align(lipgloss.Left)).Border(table.Border{}).HeaderStyle(lipgloss.NewStyle().Bold(true).Background(lipgloss.Color("#7D56F4"))).SortByAsc("name").
		WithTargetWidth(w + 4).Focused(true).WithNoPagination()
```

but kept running into problems where the progress bar text itself was being truncated prematurely.  I suspect this is due to the string containing ANSI control sequences.

Using [examples/updates/main.go](https://github.com/Evertras/bubble-table/blob/main/examples/updates/main.go) to demonstrate if I make an edit like this:

```diff
diff --git a/examples/updates/main.go b/examples/updates/main.go
index 207a9ac..ab79a78 100644
--- a/examples/updates/main.go
+++ b/examples/updates/main.go
@@ -42,7 +42,7 @@ func NewModel() Model {
 func refreshDataCmd() tea.Msg {
        // This could come from some API or something
        return []*SomeData{
-               NewSomeData("abc"),
+               NewSomeData("\033[00;37mabc"),
                NewSomeData("def"),
                NewSomeData("123"),
                NewSomeData("ok"),
```

`go run ./examples/updates` renders something like this:

<img width="244" alt="image" src="https://user-images.githubusercontent.com/324373/166063363-647da266-3076-4201-8630-2529de138648.png">

while the following code change:

```diff
diff --git a/examples/updates/main.go b/examples/updates/main.go
index 207a9ac..9a23628 100644
--- a/examples/updates/main.go
+++ b/examples/updates/main.go
@@ -42,7 +42,7 @@ func NewModel() Model {
 func refreshDataCmd() tea.Msg {
        // This could come from some API or something
        return []*SomeData{
-               NewSomeData("abc"),
+               NewSomeData("\033[00;37m\033[00;37mabc"),
                NewSomeData("def"),
                NewSomeData("123"),
                NewSomeData("ok"),
```

renders like this when you run `go run ./examples/updates`:

<img width="267" alt="image" src="https://user-images.githubusercontent.com/324373/166063276-edbde731-9124-4d46-976b-ad01b963841b.png">

It _looks_ like this is a result of how code in `table/strlimit.go` calculates the length of a string being considered for truncation.

https://github.com/Evertras/bubble-table/blob/3e8eabee4ad1dbb0e006475fcb23f1d5c890ca51/table/strlimit.go#L19-L21

If I make the following change

```diff
diff --git a/table/strlimit.go b/table/strlimit.go
index e21d97e..9c374d2 100644
--- a/table/strlimit.go
+++ b/table/strlimit.go
@@ -3,7 +3,7 @@ package table
 import (
        "strings"
 
-       "github.com/mattn/go-runewidth"
+       "github.com/charmbracelet/lipgloss"
 )
 
 func limitStr(str string, maxLen int) string {
@@ -16,8 +16,8 @@ func limitStr(str string, maxLen int) string {
                str = str[:newLineIndex] + "…"
        }
 
-       if runewidth.StringWidth(str) > maxLen {
-               return runewidth.Truncate(str, maxLen-1, "") + "…"
+       if lipgloss.Width(str) > maxLen {
+               return lipgloss.NewStyle().SetString(str).MaxWidth(maxLen-1).String() + "…"
        }
 
        return str
```

my use case of embedding ANSI control codes in text that is rendered in a cell in a table works as expected.

Does this use case - allowing user supplied ANSI control characters in cell contents - make sense?

Would this be something you'd consider accepting a PR for (assuming appropriate tests are included, of course)?

**Comment by Evertras** [comment]
Hello, thanks for the feedback!  I really appreciate the detailed breakdown, it makes it super clear to understand what's going on and what you'd like it to do.

Using embedded ANSI control characters is not a use case I had considered, but I do see this as a valid use case particularly in working nicely with progress bars as in your original snippet.

I'm wondering if https://github.com/muesli/reflow/blob/master/ansi/buffer.go#L21 might get a similar result, but I see that lipgloss would then also handle a cleaner truncation rather than cutting off an ANSI control mid-step when it comes to the rerendering.  I don't have a cleaner answer than what you've posted here off the top of my head, and for this project I prefer correctness over performance.  I'd be curious to see any benchmark hits, but assuming they're not terrible then that's fine.

I would happily accept a PR for this, especially if it added some tests around ANSI control code truncation!

**Comment by Evertras** [comment]
Thanks again for the contribution!  I've released this as `v0.11.1`, feel free to close this issue if it works for you or we can figure out next steps if something is still broken.

**Comment by greglanthier** [comment]
I just ran my local tests using `v0.11.1` and it looks like embedding progress bars works exactly as intended.

<img width="212" alt="image" src="https://user-images.githubusercontent.com/324373/166260297-56998aa7-38d7-4518-9b8d-bd4861d48a9f.png">
<img width="137" alt="image" src="https://user-images.githubusercontent.com/324373/166260431-efcf8090-a37e-4053-ac26-0f331b363358.png">

Thank you for being so welcoming to my contribution and for such a terrific Golang library.
